sourcegraph/learn
# Add interactive search

### What should this PR do?
Resolves [DEVED-110](https://sourcegraph.atlassian.net/browse/DEVED-110) by adding an interactive search component, to do interactive searches in `posts` using the Sourcegraph product.

Right now, this feature will return a set number of string query matches from a locally configured Sourcegraph Docker container (configuration details below). In the short term, it is possible to iterate on this WIP using a local Docker setup. The next step will be configuring a more robust test setup on a dedicated test VM using the Compose workflow. From there, we can move on to configuring our prod setup.

### Why are we making this change?
- We would like to have interactivity on our `posts`, and to be able to demonstrate Sourcegraph in action.

### What are the acceptance criteria? 
TBD

### How should this PR be tested?
- Install [Ngrok](https://ngrok.com/) locally, if you do not already have it.
- In your `ngrok.yml` config file, add the following config to create two tunnels below your `authtoken`:
```
tunnels:
  first:
    addr: 3000
    proto: http    
  second:
    addr: 7080
    proto: http
  ```
  **NB**: You can create multiple tunnels on a free tier account
  - Create the tunnels with `./ngrok start --all`
  - Clone `sourcegraph/sourcegraph` if you do not have it already, navigate to the `dev` directory, and run the `run-server-image.sh` script: `./run-server-image.sh`
  - Navigate to `localhost:7080` and sign up for an admin account on your local instance.
  - Once you are in, generate an access token under `http://localhost:7080/site-admin/tokens`. Note: this does not need to be a `sudo` token.
  - In your site config (`http://localhost:7080/site-admin/configuration`), add the following values, using your `ngrok` output from the terminal, which should look like this:
  ```
Forwarding                    http://21dda4518c8a.ngrok.io -> http://localhost:7080                                                            
Forwarding                    https://21dda4518c8a.ngrok.io -> http://localhost:7080                                                           
Forwarding                    http://eacfc82777a1.ngrok.io -> http://localhost:3000                                                            
Forwarding                    https://eacfc82777a1.ngrok.io -> http://localhost:3000         
```
In your Sourcegraph site config, add the following values:
```
"externalURL": "<your ngrok 7080 tunnel>",
"corsOrigin": "<your ngrok 3000 tunnel>",
```
- You can now add some GitHub repos as you normally would, once you have added your GitHub token. *NB* from testing — I would not add too many — the local bind seemed to hit limits on macOS when syncing beyond a certain number of repos, which may have to do with the way bind mounts and cache clearing are handled on macOS v Linux, where this image is meant to be run.
- Fill in the following values in your local `.env` file in the `learn` directory:
```
NEXT_PUBLIC_SEARCH_API_AUTH_TOKEN='<your local sourcegraph token>'
NEXT_PUBLIC_SEARCH_API_URL='<your ngrok 7080 tunnel>/.api/graphql'
```
- Start up your dev server: `npm run dev`
- Navigate to `regular-expression-patterns` to see the `SourcegraphInteractiveSearch` component in action, or add it to another record of your choosing using the following pattern:

```
<SourcegraphInteractiveSearch initialQuery=<your query> />
```
## Pull request process

**Reviewers**:

1. Test functionality using the criteria above.
2. Offer tips for efficiency, feedback on best practices, and possible alternative approaches and things that may not have been considered.
3. For shorter, "quick" PRs, use your best judgement on #​2.
4. Use a collaborative approach and provide resources and/or context where appropriate.
5. Provide screenshots/grabs where appropriate to show findings during review.

**Reviewees**:

1. Prefer incremental and appropriately-scoped changes.
2. Leave a comment on things you want explicit feedback on.
3. Respond clearly to comments and questions.
